Improve test.py file and add new module bookshelf.

Zavioer opened the pull request:
Module bookshelf contains Bookshelf object with unit tests.

Zavioer committed:
Improve test.py file and add new module bookshelf.

Module bookshelf contains Bookshelf object with unit tests.

amasend commented on bookshelf/bookshelf_object.py:
```diff
@@ -0,0 +1,153 @@
+
+
+class Bookshelf:
+    def __init__(self, name: str, owner: str, size: int) -> None:
```
For class constructor we usually do not put docstrings, instead of putting it here you can place it directly in a class:
```python
class Bookshelf:
    """
    Your docstring for constructor and class description.
    """
```
Overall, good work with docstrings! ;)
Find your style and keep it, or if any project defines own styles, just grab it and use it :dancers:

amasend commented on bookshelf/bookshelf_object.py:
```diff
@@ -0,0 +1,153 @@
+        """
+        return self.maxSize - self.currentBookNumber
+
+    def is_empty(self) -> bool:
```
Good work with annotations, now I can clearly see what to expect by these methods.

amasend commented on bookshelf/bookshelf_object.py:
```diff
@@ -0,0 +1,153 @@
+        more than 0 elements.
+        """
+
+        if self.books.__len__() == 0:
```
Here is some critical issue.
This is probably some habit from C or C++?
Python is mostly dedicated to be readable. Usage of dunder methods as `__len__()` or `__eq__()` etc. are mostly reserved for some really internal implementations. When you need to decide what is better in implementation as `len()` or `__len__()`, 99% of the time it will be `len()`. Why? Python defines such user friendly methods just to use them :D It is more clear when you read it and more convenient for other programmers during some code modifications.

Comparing to C and C++, you have there `public` / `protected` / `private` methods or class variables.
These things are strictly defined by these languages.
Here in Python, we do not have such things. To be honest, everything is `public`. But...
There is a convention that methods or variables with one underscore suffix we treated as protected (we in the meaning of programmers and IDEs). Methods and variables with double underscore are treated as private.
So, we should stick for that convention and not break it. There is no reason to call a private `len()` method on `books`, as user also should be able to call it. There is no reason to hide it.

One possible suggestion:
```suggestion
        if len(self.books) == 0:
```

Or other, probably less readable, but you can check if dictionary is not empty:
```suggestion
        if not self.books:
```

amasend commented on bookshelf/tests_bookshelf.py:
```diff
@@ -0,0 +1,55 @@
+from bookshelf_object import Bookshelf
+
+
+class BookTestCase(unittest.TestCase):
```
Overall, test looks good! ;)

amasend commented on bookshelf/tests_bookshelf.py:
```diff
@@ -0,0 +1,55 @@
+
+
+class BookTestCase(unittest.TestCase):
+    def setUp(self) -> None:
```
Here you used setup for object testcase. I mean, setup for the object instance of the class `BookTestCase`.
Can you try to run entire test suite written by you (this file) but with one little modification? Just add one line in the end of this setup:

```ptyhon
def setUp(self) -> None:
        self.myBookshelf = Bookshelf("Favourite books", "Jan Kowalski", 10)
        self.myBookshelf.add_book("author2", "book title")
        print('setup')
```

How many times string `setup` appears?
What is the reason?
Try to add a class setup instead of the object setup:
```python
    @classmethod
    def setUpClass(cls) -> None:
        cls.myBookshelf = Bookshelf("Favourite books", "Jan Kowalski", 10)
        cls.myBookshelf.add_book("author2", "book title")
        print('class setup')
```

How many times string `class setup` appears?
Why?
Which one is better and why?

amasend reviewed (commented):
It is really nice to see this understanding from one day of learning!
I've also seen a generated doc (multiple files from sphinx).
The most common approach here is to create a separate folder/directory to put all of the sphinx configuration there eg.:
```
-main project directory
|
---- docs
         |
         - sphinx config and docs build
```

amasend reviewed (changes requested):
Please make a separate `docs` directory and put all of the sphinx generated files there.
Also, for all of the learning files like `test.py`, `bookshelf` etc. make separate directory e.g.: `introduction_and_learning` and put these files there. We need to somehow clean our repo before we can go with normal development.

In addition, I've created a `github action` (aka workflow) to automatically check your code style in python ;)
It will be triggered whenever you create a PR to `dev` branch. We can discuss what this exactly is tomorrow morning, also I can do a short introduction how to setup such a workflow.

amasend commented:
Triggering test

Zavioer committed:
Clean repo mess

amasend commented:
Please update your fork with the latest changes from dev, we will see if `pylama` will trhow any other errors (I've made a change).

Zavioer committed:
Attempt to implement programming interface

Litte attempt to use gained knowledge. I have few questions that I can write in pule request annotations.

Zavioer commented on introduction_and_learning/Histogram/histogram_api.py:
```diff
@@ -0,0 +1,90 @@
+        generate_report(histogram, output_type, path)
+    """
+
+    @abstractmethod
+    def read_data(self, file_path: str) -> str:
+        """
+        This method handling opening and reading from given file. Type of this
+        file is given in class definition.
+
+        Parameters
+        ----------
+            file_path: string, required
+                Localisation of file which data should be read from.
+
+        Returns
+        -------
+        Method return before read data on success or information
+        "File does not have [file extension like .txt] extension." on
+        failure.
+
+        Exaples
+        -------
+        >>> histogram_txt.read_data("myDoc/file.txt")
+        >>> histogram_json.read_data("myDoc/file.json")
+        """
+        raise NotImplementedError
```
Here I was a little bit confused beceause I do not where to write docstring and I decided to write it here beceause all classes and method based on this abstract class will be do same things. But in the other hand writing docstrings in inherited class could make code more readable. Which option is better? or docstring should be here in abstract class and in inherited class. 

My second question about define method body here in abstract class or I can not do this and I can only write here `raise NotImplementedError`? I ask about this beceause methods `create` and `generate_report` are the same in all three classes defined in `histograms.ph` file, or I did someting wrong?

amasend replied:
Overall, yes, you can define a method body in abstract class method and this is a common approach when you have the same implementation ;) no issues here.

About docstring, you can inherit a docstring from a parent class like this:
`child.__doc__ = parent.__doc__` for example in class constructor or even in the class body directly.

amasend commented on introduction_and_learning/Histogram/.idea/.gitignore:
```diff
@@ -0,0 +1,3 @@
+# Default ignored files
```
.idea files should be added to .gitignore as all of them are from internal Pycharm settings

amasend commented on introduction_and_learning/Histogram/.idea/Histogram.iml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Same as above.

amasend commented on introduction_and_learning/Histogram/.idea/inspectionProfiles/profiles_settings.xml:
```diff
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
```
Same as above.

amasend commented on introduction_and_learning/Histogram/.idea/misc.xml:
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Same as above.

amasend commented on introduction_and_learning/Histogram/.idea/modules.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Same as above.

amasend commented on introduction_and_learning/Histogram/histogram_api.py:
```diff
@@ -0,0 +1,90 @@
+        This method handling opening and reading from given file. Type of this
+        file is given in class definition.
+
+        Parameters
```
I am not sure if we need to describe parameters etc. here, most of the time, abstract methods should be described shortly. This is not an interface toward the user, but rather towards programmer. Probably shorter description would be better, and longer one in public classes that user will be using.

amasend commented on introduction_and_learning/Histogram/histograms.py:
```diff
@@ -0,0 +1,291 @@
+            with open(file_path, "tr", encoding="utf8") as f:
+                return f.read()
+        else:
+            return "File does not have a .txt extension."
```
I would rather raise some custom crafted error here instead of returning some string.
As User wants to use `HistogramTxt.read_data()` on non `.txt` file, we should clearly tell him/her that this is illegal. Do you agree?
Anyway, we should use exceptions in situations when it is something wrong, this will be more readable and better exposed to the user.

amasend commented on introduction_and_learning/Histogram/histograms.py:
```diff
@@ -0,0 +1,291 @@
+        # --- end note
+        hist = {}
+
+        for word in words:
```
Tip: try to use a built in `Counter` class for that:
`from collections import Counter`
https://docs.python.org/2/library/collections.html#collections.Counter

amasend commented on introduction_and_learning/Histogram/histograms.py:
```diff
@@ -0,0 +1,291 @@
+        return hist
+
+    def generate_report(self, histogram: dict, output_type: str = "html", path: str = "report") -> bool:
+        output_types = ["html", "csv", "txt"]
```
You can define some Enums for that:
e.g.:
by making a simple class
```python
class FileTypes:
    HTML = 'html'
    ...
```

or just use an `Enum` class from `enum` module

amasend commented on introduction_and_learning/Histogram/histograms.py:
```diff
@@ -0,0 +1,291 @@
+                output = ["<html>\n<body>\n\t<table>\n\t<tr><th>Name</th>"
+                          "<th>Amount</th></tr>"]
+
+                for key in histogram.keys():
```
To get keys and values at the same time you can use:
`for key, value in histogram.items():`

amasend commented on introduction_and_learning/Histogram/histograms.py:
```diff
@@ -0,0 +1,291 @@
+                    x = "".join(output)
+
+                    if f.write(x) > 0:
+                        f.close()
```
When you use a `with` statement, you should not bother about closing this context (in this example a file).
```
with locked(myLock):
    # Code here executes with myLock held.  The lock is
    # guaranteed to be released when the block is left (even
    # if via return or by an uncaught exception).
```
taken from: https://www.python.org/dev/peps/pep-0343/

amasend reviewed (dismissed):
Yeah, overall good job.
We can see an understanding about interface. I will approve and you can merge it to dev, do not bother about lintern for this one.

Zavioer committed:
SIR model first step

Attempt to create abstract class for SIR people groups and simple demo of  how this will be working in my idea. After weekend I try to upgrade PopulationGroup class.

amasend commented:
Can you make other PR for simulation implementation? It will be better to distinguish this work from the preparation work. Just remove these commits from your dev branch, and create eg. simulation branch and make PR from your simulation towards this dev ok?

Zavioer commented:
I created new PR.
